src/login/login.cpp

```diff
@@ -464,6 +491,62 @@ bool login_check_password(const char* md5key, int passwdenc, const char* passwd,
 	}
 }
 
+/**
+ * Check if password needs to be updated, and update it if necessary.
+ * Assume password is already validated as correct
+ * @param passwd: raw passwd
+ * @param acc: the account to update
+ * @return 1 if updated, 0 if already updated, -1 if error
+ */
+int login_update_password(const char * passwd, struct mmo_account *acc) {
+	if (acc->passwd_type == 1) { // already done
+		return 0;
+	}
+	try {
+		// not hashed correctly (just bcrypt), update acc with new hash
+		auto dbpasslen = strlen(acc->pass);
+		bool was_md5 = (dbpasslen == 32); // MD5 is 32 chars
+		safestrncpy(acc->pass, BCrypt::generateHash(passwd).c_str(), sizeof(acc->pass));
+		acc->passwd_type = was_md5 ? 2 : 1;
+	} catch (const std::exception& e) {
+		ShowDebug("%s\n", e.what());
+		return -1;
+	}
+	ShowInfo("Encrypted password for %s\n", acc->userid);
+	return 1;
+}
+
+/**
+ * Update all passwords.
+ * Stop the login server after this is done.
+ */
+void login_update_all_passwords() {
+
```

Comment:
Instead of duplicating code here again I would prefer to simply add some kind of state tracking that launches this after "normal" initialization.
Then you also dont need those hacks below the loop.